pkg/policy/api/utils.go

```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+	// This will generate one-address CIDRs consisting of endpoint backend ip
+	mask := net.CIDRMask(128, 128)
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+
+		found := false
+		for _, c := range egress.ToCIDR {
+			_, cidr, err := net.ParseCIDR(string(c))
+			if err != nil {
+				return err
+			}
+			if cidr.Contains(epIP) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, CIDR(cidr.String()))
+		}
+	}
+	return nil
+}
+
+// generateToPortsFromEndpoint takes an egress rule and populates it with ToPorts rules based on provided enpoint object
+func generateToPortsFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
+	// additional port rule that will contain all endpoint ports
+	portRule := PortRule{}
+	for _, port := range endpoint.Ports {
+		found := false
+	loop:
+		for _, portRule := range egress.ToPorts {
+			for _, portProtocol := range portRule.Ports {
+				numericPort, err := strconv.Atoi(portProtocol.Port)
+				if err != nil {
+					return err
+				}
+
+				if strings.ToLower(string(port.Protocol)) == strings.ToLower(string(portProtocol.Protocol)) && int(port.Port) == numericPort {
+					found = true
+					break loop
+				}
+			}
+		}
+		if !found {
+			portRule.Ports = append(portRule.Ports, PortProtocol{
+				Port:     strconv.Itoa(int(port.Port)),
+				Protocol: L4Proto(strings.ToUpper(string(port.Protocol))),
+			})
+		}
+	}
+
+	if len(portRule.Ports) > 0 {
+		egress.ToPorts = append(egress.ToPorts, portRule)
+	}
+
+	return nil
+}
+
+// DeleteGeneratedToServiceRulesFromEndpoint removes all ToService-based ToCIDR
+// rules generated from endpoint
+func (e *EgressRule) DeleteGeneratedToServiceRulesFromEndpoint(serviceInfo types.K8sServiceNamespace, endpoint types.K8sServiceEndpoint) error {
+	for _, service := range e.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == serviceInfo {
+			if err := deleteToCidrFromEndpoint(e, endpoint); err != nil {
+				return err
+			}
+			// TODO: deleteToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+// deleteToCidrFromEndpoint takes an egress rule and removes ToCIDR rules matching endpoint
+func deleteToCidrFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
+	newToCIDR := make([]CIDR, 0, len(egress.ToCIDR))
+
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+		for _, c := range egress.ToCIDR {
+			_, cidr, err := net.ParseCIDR(string(c))
+			if err != nil {
+				return err
+			}
+			if !cidr.Contains(epIP) {
+				//if endpoint is not in CIDR it's ok to retain it
+				newToCIDR = append(newToCIDR, c)
+			}
+		}
+	}
+
+	egress.ToCIDR = newToCIDR
+
+	return nil
+}
+
+// deleteToPortsFromEndpoint takes an egress rule and removes ToPorts rules matching endpoint
+func deleteToPortsFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
```

Comment:
Not sure if relevant, but is any locking required while deleting /updating endpoint ?

Reply:
Locking is done on repository level ATM, https://github.com/cilium/cilium/pull/1785/files#diff-7c38bf553d4d2690add68b2464b58ef0R480

Reply:
This function is not used